client/files_test.go

```diff
@@ -0,0 +1,194 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client_test
+
+import (
+	"encoding/json"
+	"os"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/canonical/pebble/client"
+)
+
+type makeDirPayload struct {
+	Action string         `json:"action"`
+	Dirs   []makeDirsItem `json:"dirs"`
+}
+
+type makeDirsItem struct {
+	Path        string `json:"path"`
+	MakeParents bool   `json:"make-parents"`
+	Permissions string `json:"permissions"`
+	UserID      *int   `json:"user-id"`
+	User        string `json:"user"`
+	GroupID     *int   `json:"group-id"`
+	Group       string `json:"group"`
+}
+
+func (cs *clientSuite) TestMakeDir(c *C) {
+	cs.rsp = `{"type": "sync", "result": [{"path": "/foo/bar"}]}`
+
+	err := cs.cli.MakeDir(&client.MakeDirOptions{
+		Path:        "/foo/bar",
+		MakeParents: true,
+	})
+	c.Assert(err, IsNil)
+
+	c.Assert(cs.req.URL.Path, Equals, "/v1/files")
+	c.Assert(cs.req.Method, Equals, "POST")
+
+	var payload makeDirPayload
+	decoder := json.NewDecoder(cs.req.Body)
+	err = decoder.Decode(&payload)
+	c.Assert(err, IsNil)
+	c.Check(payload, DeepEquals, makeDirPayload{
+		Action: "make-dirs",
+		Dirs: []makeDirsItem{{
+			Path:        "/foo/bar",
+			MakeParents: true,
+		}},
+	})
+}
+
+func (cs *clientSuite) TestMakeDirWithPermissions(c *C) {
+	cs.rsp = `{"type": "sync", "result": [{"path": "/foo/bar"}]}`
+
+	permissions := os.FileMode(0o755)
+	err := cs.cli.MakeDir(&client.MakeDirOptions{
+		Path:        "/foo/bar",
+		MakeParents: true,
+		Permissions: &permissions,
+	})
+	c.Assert(err, IsNil)
+
+	c.Assert(cs.req.URL.Path, Equals, "/v1/files")
+	c.Assert(cs.req.Method, Equals, "POST")
+
+	var payload makeDirPayload
+	decoder := json.NewDecoder(cs.req.Body)
+	err = decoder.Decode(&payload)
+	c.Assert(err, IsNil)
+	c.Check(payload, DeepEquals, makeDirPayload{
+		Action: "make-dirs",
+		Dirs: []makeDirsItem{{
+			Path:        "/foo/bar",
+			MakeParents: true,
+			Permissions: "755",
+		}},
+	})
+}
+
+func (cs *clientSuite) TestMakeDirFails(c *C) {
+	cs.rsp = `{"type": "error", "result": {"message": "could not foo"}}`
+	err := cs.cli.MakeDir(&client.MakeDirOptions{
+		Path: "/foobar",
+	})
+	c.Assert(err, ErrorMatches, "could not foo")
+
+	c.Assert(cs.req.URL.Path, Equals, "/v1/files")
+	c.Assert(cs.req.Method, Equals, "POST")
+
+	var payload makeDirPayload
+	decoder := json.NewDecoder(cs.req.Body)
+	err = decoder.Decode(&payload)
+	c.Assert(err, IsNil)
+	c.Check(payload, DeepEquals, makeDirPayload{
+		Action: "make-dirs",
+		Dirs: []makeDirsItem{{
+			Path: "/foobar",
+		}},
+	})
+}
+
+func (cs *clientSuite) TestMakeDirFailsOnDirectory(c *C) {
+	cs.rsp = `
+{
+	"type": "sync",
+	"result": [
+		{
+			"path": "/foobar",
+			"error": {
+				"message": "could not bar",
+				"kind": "permission-denied",
+				"value": 42
+			}
+		}
+	]
+}`
+
+	err := cs.cli.MakeDir(&client.MakeDirOptions{
+		Path: "/foobar",
+	})
+	c.Assert(err, ErrorMatches, "could not bar")
```

Comment:
Here's where we would test that `err` type-asserts to `*client.Error` with `Kind == "permission-denied"`.